pkg/cmd/cli/cmd/logs.go

```diff
@@ -43,204 +40,129 @@ If the pod has only one container, the container name is optional.`
-	cmd.Flags().BoolVarP(&o.KubeLogOptions.Previous, "previous", "p", o.KubeLogOptions.Previous, "If true, print the logs for the previous instance of the container in a pod if it exists.")
-	cmd.Flags().IntVar(&o.KubeLogOptions.Tail, "tail", o.KubeLogOptions.Tail, "Lines of recent log file to display. Defaults to -1, showing all log lines.")
-	cmd.Flags().String("since-time", "", "Only return logs after a specific date (RFC3339). Defaults to all logs. Only one of since-time / since may be used.")
-	cmd.Flags().DurationVar(&o.KubeLogOptions.SinceSeconds, "since", o.KubeLogOptions.SinceSeconds, "Only return logs newer than a relative duration like 5s, 2m, or 3h. Defaults to all logs. Only one of since-time / since may be used.")
-	cmd.Flags().StringVarP(&o.KubeLogOptions.ContainerName, "container", "c", o.KubeLogOptions.ContainerName, "Container name")
-
 	return cmd
 }
 
 func (o *OpenShiftLogsOptions) Complete(f *clientcmd.Factory, out io.Writer, cmd *cobra.Command, args []string) error {
-	switch len(args) {
-	case 0:
-		return cmdutil.UsageError(cmd, "RESOURCE is required for log")
-
-	case 1:
-		o.ResourceString = args[0]
-	case 2:
-		o.ResourceString = args[0]
-		o.KubeLogOptions.ContainerName = args[1]
-
-	default:
-		return cmdutil.UsageError(cmd, "log RESOURCE")
-	}
-
-	var err error
-	o.Namespace, _, err = f.DefaultNamespace()
-	if err != nil {
+	if err := o.KubeLogOptions.Complete(f.Factory, out, cmd, args); err != nil {
 		return err
 	}
-	o.OriginClient, o.KubeLogOptions.Client, err = f.Clients()
+	namespace, _, err := f.DefaultNamespace()
 	if err != nil {
 		return err
 	}
 
-	o.KubeLogOptions.PodName = o.ResourceString
-	o.KubeLogOptions.PodNamespace = o.Namespace
+	podLogOptions := o.KubeLogOptions.Options.(*kapi.PodLogOptions)
 
-	sinceTime := cmdutil.GetFlagString(cmd, "since-time")
-	if len(sinceTime) > 0 {
-		t, err := kapi.ParseRFC3339(sinceTime, unversioned.Now)
-		if err != nil {
-			return err
-		}
-		o.KubeLogOptions.SinceTime = &t
-	}
-
-	return nil
-}
-
-func (o *OpenShiftLogsOptions) Validate() error {
-	if len(o.ResourceString) == 0 {
-		return errors.New("RESOURCE must be specified")
-	}
-
-	return o.KubeLogOptions.Validate()
-}
-
-func (o *OpenShiftLogsOptions) RunLog() error {
-	resourceType := "pod"
-	resourceName := o.ResourceString
-	tokens := strings.SplitN(o.ResourceString, "/", 2)
-	if len(tokens) == 2 {
-		resourceType = tokens[0]
-		resourceName = tokens[1]
-	}
-	resourceType = strings.ToLower(resourceType)
-
-	// if we're requesting a pod, delegate directly to kubectl logs
-	if (resourceType == "pods") || (resourceType == "pod") || (resourceType == "po") {
-		o.KubeLogOptions.PodName = resourceName
-		return o.KubeLogOptions.RunLog()
-	}
-
-	if len(o.KubeLogOptions.ContainerName) > 0 {
-		return errors.New("container cannot be specified with anything besides a pod")
+	mapper, typer := f.Object()
+	infos, err := resource.NewBuilder(mapper, typer, f.ClientMapperForCommand()).
```

Comment:
Just so I'm sure I'm following intent. All you actually want here is the type of resource being requested, so that you build the correction type of `LogOptions` object.  There's no real reason you're making a server request other than this method is convenient, right?